modules/memfault-firmware-sdk/memfault_gnss_metrics.c

```diff
@@ -0,0 +1,141 @@
+static void memfault_gnss_metrics_stop_fix_session(const struct location_event_data *event_data);
+
+// To maintain separation of metrics collection from NCS library code, we need a way to
+// know when a fix request is made, and when the result of the request is ready (fix or error).
+// There is a location event handler that indicates a location request result, but only one,
+// which should be reserved for application code. There is also no callback mechanism to
+// be notified when a fix request is made. Therefore, we wrap the functions of interest:
+//        - location_init() -- registers an event callback. Save the application
+//              callback and register a wrapper callback.
+//        - location_request() -- initiates a request. Intercept it's result, and
+//              start a session if the request is successful.
+int __real_location_init(location_event_handler_t event_handler);
+int __wrap_location_init(location_event_handler_t event_handler)
+{
+	real_location_event_handler = event_handler;
+	return __real_location_init(memfault_location_event_handler);
+}
+
+// Ensure the substituted function signature matches the original function
+_Static_assert(__builtin_types_compatible_p(__typeof__(&location_init),
+					    __typeof__(&__wrap_location_init)) &&
+		       __builtin_types_compatible_p(__typeof__(&location_init),
+						    __typeof__(&__real_location_init)),
+	       "Error: Wrapped function does not match original location_init function signature");
+
+// Also wrap the location request to know when to start a session
+int __real_location_request(const struct location_config *config);
+int __wrap_location_request(const struct location_config *config)
+{
+	int result = __real_location_request(config);
+
+	if (result == 0) {
+		memfault_gnss_metrics_start_fix_session();
+	}
+	return result;
+}
+
+// Ensure the substituted function signature matches the original function
+_Static_assert(
+	__builtin_types_compatible_p(__typeof__(&location_request),
+				     __typeof__(&__wrap_location_request)) &&
+		__builtin_types_compatible_p(__typeof__(&location_request),
+					     __typeof__(&__real_location_request)),
+	"Error: Wrapped function does not match original location_request function signature");
+
+static void memfault_location_event_handler(const struct location_event_data *event_data)
+{
+	if (event_data->id == LOCATION_EVT_LOCATION || event_data->id == LOCATION_EVT_TIMEOUT ||
+	    event_data->id == LOCATION_EVT_ERROR || event_data->id == LOCATION_EVT_RESULT_UNKNOWN) {
+		memfault_gnss_metrics_stop_fix_session(event_data);
+	}
+	real_location_event_handler(event_data);
+}
+
+/**
+ * @brief Start a GNSS fix session.
+ *
+ */
+static void memfault_gnss_metrics_start_fix_session(void)
+{
+	if (session_in_progress) {
+		return;
+	}
+
+	session_in_progress = true;
+	LOG_DBG("Starting GNSS session");
+	MEMFAULT_METRICS_SESSION_START(ncs_gnss);
+	MEMFAULT_METRIC_ADD(ncs_gnss_fix_request_count, 1);
```

Comment:
The metrics talk about GNSS. I don't know if this is intentional but the location request may be done with other methods too and there might be fallbacks to other methods. The default order of methods is GNSS, Wi-Fi, cellular if all are supported by HW and the application. If GNSS fails, Location library will try Wi-Fi and cellular. Currently this PR is creating GNSS metrics without knowing which method is actually used.

Reply:
Good point. I should instead call these metrics `ncs_location_*` to be able to capture location metrics from systems that use more than the GNSS methods.

Reply:
I think we should think about what data we actually want to collect. If we collect succeeded and failed location requests that may not provide much data for debugging as you don't really know which methods have succeeded and failed.

Reply:
Earlier implementation within Asset Tracker v2 collected only GNSS fix times and not cellular related things.
On the other hand, timeout event handling was there to get timeouts from all methods and not just GNSs so it was not really consistent.

Reply:
> If we collect succeeded and failed location requests that may not provide much data for debugging as you don't really know which methods have succeeded and failed.

Fair point. We are definitely wanting to collect GNSS-specific data as mentioned in another thread. I'm less familiar with the WiFi and cellular methods, but I'll poke around if you think you'd be open to adding more details from those methods to the location data passed to the handler.